### PR TITLE
test(typeutils): add datatype.go unit test coverage

### LINKED ISSUE #644: feat: Add Unit Tests for datatype.go
#### **Problem**

The OLake codebase currently has unit tests for:

* `resolver`
* `flattener`
* `catalog`

However, the file `olake/utils/typeutils/datatype.go` does not have any test coverage.

This file handles critical type detection and comparison logic:
* `TypeFromValue` 
* `MaximumOnDataType`
* `detectTimestampPrecision`
* `ExtractAndMapColumnType`

Without unit tests, bugs in type detection, timestamp precision handling, or type comparisons may slip through during development.

---

#### **Proposed Solution**

Add comprehensive unit tests for the logic inside `datatype.go`.  
This will help:

* Increase overall test coverage
* Improve type detection reliability
* Catch bugs earlier during CI
* Make contributions safer

For reference on how existing unit tests are structured, you can refer to **PR #273**.

---

#### **Notes for Contributors**

* Add test covering most of the scenarios and edge cases which should indicate both pass and fail scenarios.
* Follow the existing test file structure and naming conventions to maintain the consistency with other part.

### PR DESCRIPTION
**Pull request by vimla01**
# Description

Added comprehensive unit tests for `utils/typeutils/datatype.go` to cover core type detection and comparison logic:

- `TypeFromValue`
- `MaximumOnDataType`
- `detectTimestampPrecision`
- `ExtractAndMapColumnType`

This includes normal and edge/error scenarios: nils, pointers/double pointers, reflect values, json.Number variants, timestamp precision paths, unsupported datatype comparisons, and SQL type mapping cases (`bigint`, `float`, `double`, `timestamp`, `unknown`, empty/space input).

Also fixed one behavior gap in `datatype.go`: invalid `reflect.Value` now maps to `types.Null`, matching expected handling for invalid/reflect-invalid input.

Fixes #644

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

- [x] Scenario A: Focused datatype tests
  - `go test -count=1 -v ./utils/typeutils -run 'TestTypeFromValue|TestMaximumOnDataTypeTimestamp|TestMaximumOnDataTypeInt64|TestMaximumOnDataTypeUnsupported|TestDetectTimestampPrecision|TestExtractAndMapColumnType'`
- [x] Scenario B: Full package verification + coverage
  - `go test -count=1 ./utils/typeutils`
  - `go test -count=1 ./utils/typeutils -coverprofile=/tmp/typeutils.out`
  - `go tool cover -func=/tmp/typeutils.out | grep 'utils/typeutils/datatype.go'`
  - `go test -count=1 ./utils/typeutils -cover`

## Recording :
[test.webm](https://github.com/user-attachments/assets/49c136b1-e091-4f15-836d-79a39644d518)

## Documentation

- [ ] Documentation Link: [link to README, olake.io/docs, or olake-docs]
- [x] N/A (bug fix, refactor, or test changes only)

## Related PR's (If Any)
None

### CODE REVIEW

**Comment by nayanj98** [comment]
@vimla01 If you can provide a video of you running the sync using OLake that would be helpful as it is primary requirement. Like you can make a basic video showing your data in source, running sync using OLake and then showing data in destination, that's it! As soon as you are able to provide that I will assign someone internally to review your PR.

**Comment by vimla01** [comment]
> @vimla01 If you can provide a video of you running the sync using OLake that would be helpful as it is primary requirement. Like you can make a basic video showing your data in source, running sync using OLake and then showing data in destination, that's it! As soon as you are able to provide that I will assign someone internally to review your PR.

Hi @nayanj98 
Uploaded the OLake sync demo video.

[olake_sync.webm](https://github.com/user-attachments/assets/e0504c17-5599-4a5d-b59f-9733bf012b6f)

**Comment by nayanj98** [comment]
@vimla01 The video looks good. I will try to assign someone soon to review your PR.

**Comment by nayanj98** [comment]
@vimla01 Assigning @vishalm0509 to review your PR

File: utils/typeutils/datatype.go [hunk 1]
```diff
@@ -94,6 +94,13 @@ func TypeFromValue(v interface{}) types.DataType {
 
 // typeFromValueReflect handles types that require reflection
 func typeFromValueReflect(v interface{}) types.DataType {
+	if rv, ok := v.(reflect.Value); ok {
+		if !rv.IsValid() {
+			return types.Null
+		}
+		return TypeFromValue(rv.Interface())
+	}
```
Comment (vishalm0509) [inline utils/typeutils/datatype.go]: Can you think of a case where this will be executed ?

File: utils/typeutils/datatype_test.go [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+
+		{name: "float32_positive", input: float32(3.14), expected: types.Float32},
+		{name: "float32_zero", input: float32(0), expected: types.Float32},
+		{name: "float32_negative", input: float32(-1.5), expected: types.Float32},
+		{name: "float64_positive", input: float64(2.718281828), expected: types.Float64},
+		{name: "float64_zero", input: float64(0), expected: types.Float64},
+
+		{name: "string_regular", input: "hello world", expected: types.String},
+		{name: "string_empty", input: "", expected: types.String},
+		{name: "string_not_date", input: "not a date", expected: types.String},
```
Comment (vishalm0509) [inline utils/typeutils/datatype_test.go]: string regular handles this case

File: utils/typeutils/datatype_test.go [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+		t.Run(tc.name, func(t *testing.T) {
+			assert.Equal(t, tc.expected, detectTimestampPrecision(tc.input))
+		})
+	}
+
+	t.Run("time_now_precision", func(t *testing.T) {
+		result := detectTimestampPrecision(time.Now())
+		assert.Contains(t, []types.DataType{types.Timestamp, types.TimestampMilli, types.TimestampMicro, types.TimestampNano}, result)
+	})
+}
```
Comment (vishalm0509) [inline utils/typeutils/datatype_test.go]: What behaviour does this test. Either of these will always be returned.

File: utils/typeutils/datatype.go (hunk 1 above)
Comment (vishalm0509) [reply utils/typeutils/datatype.go]: We can remove this check if it has been added for the test's purpose only

File: utils/typeutils/datatype_test.go [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+		{name: "time_second_precision", input: time.Date(2024, 12, 18, 10, 30, 0, 0, time.UTC), expected: types.Timestamp},
+		{name: "time_milli_precision", input: time.Date(2024, 12, 18, 10, 30, 0, 123000000, time.UTC), expected: types.TimestampMilli},
+		{name: "time_micro_precision", input: time.Date(2024, 12, 18, 10, 30, 0, 123456000, time.UTC), expected: types.TimestampMicro},
+		{name: "time_nano_precision", input: time.Date(2024, 12, 18, 10, 30, 0, 123456789, time.UTC), expected: types.TimestampNano},
+		{name: "time_zero_value", input: time.Time{}, expected: types.Timestamp},
+
+		{name: "int64_pointer", input: &int64Value, expected: types.Int64},
+		{name: "float64_pointer", input: &float64Value, expected: types.Float64},
+		{name: "double_pointer_int64", input: doublePtrInt64, expected: types.Int64},
+		{name: "string_pointer", input: &stringValue, expected: types.String},
```
Comment (vishalm0509) [inline utils/typeutils/datatype_test.go:97]: Add string pointer to time.Time datatype as well. 

```go
	case *string:
		if val == nil {
			return types.Null
		}
		t, err := ReformatDate(*val, false)
		if err == nil {
			return detectTimestampPrecision(t)
		}
		return types.String

```

--- 

also think about all the edge cases. Happy-cases, edge-cases, success/error cases, null cases for all the unit tests. If you think you have covered all then fine.

Comment (vimla01) [reply utils/typeutils/datatype_test.go:97]: Added `string_pointer_timestamp_milli` to cover *string parsing into timestamp precision.

File: utils/typeutils/datatype_test.go (hunk 3 above)
Comment (vimla01) [reply utils/typeutils/datatype_test.go]: Agreed. removed this test since it didn’t validate specific behavior

File: utils/typeutils/datatype.go (hunk 1 above)
Comment (vimla01) [reply utils/typeutils/datatype.go]: Yeah okay, I’ve removed thiss

File: utils/typeutils/datatype_test.go (hunk 2 above)
Comment (vimla01) [reply utils/typeutils/datatype_test.go]: I ve removed the redundant `string_not_date` case

File: utils/typeutils/datatype_test.go (hunk 4 above)
Comment (vimla01) [reply utils/typeutils/datatype_test.go:97]: I checked again using the checklist and covered happy-path, edge, success/error, and null scenarios across all four functions

File: utils/typeutils/datatype_test.go [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+	testCases := []struct {
+		name          string
+		leftArgument  string
+		rightArgument string
+		expected      string
+		expectError   bool
+	}{
+		{name: "left_is_max", leftArgument: "2024-12-18", rightArgument: "2024-12-17", expected: "2024-12-18"},
+		{name: "right_is_max", leftArgument: "2024-01-01", rightArgument: "2024-12-31", expected: "2024-12-31"},
+		{name: "equal_values_returns_left", leftArgument: "2024-12-18", rightArgument: "2024-12-18", expected: "2024-12-18"},
```
Comment (vishalm0509) [inline utils/typeutils/datatype_test.go:138]: It handles mostly `date` datatype. Please also add tests with full datetime values (hour/min/sec, timezone offset, milliseconds/microseconds).

For example:
```go
		{name: "datetime_second_precision_right_is_max", leftArgument: "2024-12-18T10:30:00Z", rightArgument: "2024-12-18T10:30:01Z", expected: "2024-12-18T10:30:01Z"},
		{name: "datetime_timezone_offset_left_is_max", leftArgument: "2024-12-18T10:30:00+05:30", rightArgument: "2024-12-18T04:59:59Z", expected: "2024-12-18T10:30:00+05:30"},
		{name: "datetime_millisecond_right_is_max", leftArgument: "2024-12-18T10:30:00.122Z", rightArgument: "2024-12-18T10:30:00.123Z", expected: "2024-12-18T10:30:00.123Z"},
		{name: "datetime_microsecond_left_is_max", leftArgument: "2024-12-18T10:30:00.123457Z", rightArgument: "2024-12-18T10:30:00.123456Z", expected: "2024-12-18T10:30:00.123457Z"},

        {name: "date_vs_millisecond_datetime_right_is_max", leftArgument: "2024-12-18", rightArgument: "2024-12-18T00:00:00.001Z", expected: "2024-12-18T00:00:00.001Z"},
		{name: "millisecond_datetime_vs_date_left_is_max", leftArgument: "2024-12-18T23:59:59.999Z", rightArgument: "2024-12-18", expected: "2024-12-18T23:59:59.999Z"},
```

File: utils/typeutils/datatype_test.go [hunk 6]
```diff
@@ -0,0 +1,287 @@
...
+	doublePtrInt64 := &ptrInt64
+
+	testCases := []struct {
+		name     string
+		input    any
+		expected types.DataType
+	}{
+		{name: "nil", input: nil, expected: types.Null},
+		{name: "nil_int_pointer", input: (*int)(nil), expected: types.Null},
+		{name: "nil_string_pointer", input: (*string)(nil), expected: types.Null},
```
Comment (vishalm0509) [inline utils/typeutils/datatype_test.go:40]:
```suggestion
		{name: "nil", input: nil, expected: types.Null},
		{name: "nil_int_pointer", input: (*int)(nil), expected: types.Null},
		{name: "nil_string_pointer", input: (*string)(nil), expected: types.Null},
		{name: "nil_bool_pointer", input: (*bool)(nil), expected: types.Null},
		{name: "nil_float64_pointer", input: (*float64)(nil), expected: types.Null},
		{name: "nil_time_pointer", input: (*time.Time)(nil), expected: types.Null},
```

File: utils/typeutils/datatype_test.go [hunk 7]
```diff
@@ -0,0 +1,287 @@
...
+	}
+
+	t.Run("equal_timestamps_return_left_value", func(t *testing.T) {
+		left := time.Date(2024, 12, 18, 0, 0, 0, 0, time.UTC)
+		right := time.Date(2024, 12, 18, 0, 0, 0, 0, time.UTC)
+
+		got, err := MaximumOnDataType(types.Timestamp, &left, &right)
+		require.NoError(t, err)
+		assert.Equal(t, &left, got)
+	})
```
Comment (vishalm0509) [inline utils/typeutils/datatype_test.go]: Check if we need to cover these kind of cases
```suggestion
	t.Run("nil_left_time_pointer_returns_error_and_left", func(t *testing.T) {
		right := time.Date(2024, 12, 18, 0, 0, 0, 0, time.UTC)
		var left *time.Time

		got, err := MaximumOnDataType(types.Timestamp, left, &right)
		require.Error(t, err)
		assert.Contains(t, err.Error(), "failed to reformat")
		assert.Nil(t, got)
	})

	t.Run("nil_right_time_pointer_returns_error_and_left", func(t *testing.T) {
		left := time.Date(2024, 12, 18, 0, 0, 0, 0, time.UTC)
		var right *time.Time

		got, err := MaximumOnDataType(types.Timestamp, &left, right)
		require.Error(t, err)
		assert.Contains(t, err.Error(), "failed to reformat")
		assert.Equal(t, &left, got)
	})
```

File: utils/typeutils/datatype_test.go [hunk 8]
```diff
@@ -0,0 +1,287 @@
...
+}
```
Comment (vishalm0509) [inline utils/typeutils/datatype_test.go:166]:
```suggestion
func TestMaximumOnDataTypeTimestamp(t *testing.T) {
	leftPointer := time.Date(2024, 12, 18, 0, 0, 0, 0, time.UTC)
	rightPointer := time.Date(2024, 12, 18, 0, 0, 0, 0, time.UTC)
	nilLeftPointer := (*time.Time)(nil)
	nilRightPointer := (*time.Time)(nil)
	rightValue := time.Date(2024, 12, 18, 0, 0, 0, 0, time.UTC)
	leftValue := time.Date(2024, 12, 18, 0, 0, 0, 0, time.UTC)

	testCases := []struct {
		name          string
		leftArgument  any
		rightArgument any
		expected      any
		expectError   bool
	}{
		{name: "left_is_max", leftArgument: "2024-12-18", rightArgument: "2024-12-17", expected: "2024-12-18"},
		{name: "right_is_max", leftArgument: "2024-01-01", rightArgument: "2024-12-31", expected: "2024-12-31"},
		{name: "equal_values_returns_left", leftArgument: "2024-12-18", rightArgument: "2024-12-18", expected: "2024-12-18"},
		{name: "datetime_second_precision_right_is_max", leftArgument: "2024-12-18T10:30:00Z", rightArgument: "2024-12-18T10:30:01Z", expected: "2024-12-18T10:30:01Z"},
		{name: "datetime_timezone_offset_left_is_max", leftArgument: "2024-12-18T10:30:00+05:30", rightArgument: "2024-12-18T04:59:59Z", expected: "2024-12-18T10:30:00+05:30"},
		{name: "datetime_millisecond_right_is_max", leftArgument: "2024-12-18T10:30:00.122Z", rightArgument: "2024-12-18T10:30:00.123Z", expected: "2024-12-18T10:30:00.123Z"},
		{name: "datetime_microsecond_left_is_max", leftArgument: "2024-12-18T10:30:00.123457Z", rightArgument: "2024-12-18T10:30:00.123456Z", expected: "2024-12-18T10:30:00.123457Z"},
		{name: "date_vs_millisecond_datetime_right_is_max", leftArgument: "2024-12-18", rightArgument: "2024-12-18T00:00:00.001Z", expected: "2024-12-18T00:00:00.001Z"},
		{name: "millisecond_datetime_vs_date_left_is_max", leftArgument: "2024-12-18T23:59:59.999Z", rightArgument: "2024-12-18", expected: "2024-12-18T23:59:59.999Z"},
		{name: "equal_timestamps_return_left_value", leftArgument: &leftPointer, rightArgument: &rightPointer, expected: &leftPointer},
		{name: "invalid_left_value", leftArgument: "invalid-date", rightArgument: "2024-12-18", expectError: true},
		{name: "invalid_right_value", leftArgument: "2024-12-18", rightArgument: "invalid-date", expectError: true},
		{name: "nil_left_time_pointer_returns_error_and_left", leftArgument: nilLeftPointer, rightArgument: &rightValue, expectError: true},
		{name: "nil_right_time_pointer_returns_error_and_left", leftArgument: &leftValue, rightArgument: nilRightPointer, expectError: true},
	}

	for _, tc := range testCases {
		t.Run(tc.name, func(t *testing.T) {
			got, err := MaximumOnDataType(types.Timestamp, tc.leftArgument, tc.rightArgument)
			if tc.expectError {
				require.Error(t, err)
				assert.Contains(t, err.Error(), "failed to reformat")
				assert.Equal(t, tc.leftArgument, got)
				return
			}

			require.NoError(t, err)
			assert.Equal(t, tc.expected, got)
		})
	}
}
```

Comment (vishalm0509) [reply utils/typeutils/datatype_test.go:166]: check the new test cases added, and think if we need them and if yes, please check if we need to handle such cases in other places as well

File: utils/typeutils/datatype_test.go (hunk 5 above)
Comment (vimla01) [reply utils/typeutils/datatype_test.go:138]: added these cases in `TestMaximumOnDataTypeTimestamp`:
- full datetime comparisons with second precision
- timezone offset comparisons
- millisecond and microsecond comparisons
- mixed date vs datetime comparisons
- nil `*time.Time` error cases

File: utils/typeutils/datatype_test.go (hunk 8 above)
Comment (vimla01) [reply utils/typeutils/datatype_test.go:166]: I checked the added cases against the current implementation before keeping them. These datetime, timezone-offset, mixed date/datetime, and nil *time.Time cases are already supported through `MaximumOnDataType(types.Timestamp) `calling `ReformatDate`, so I kept them as tests and did not add extra production changes elsewhere.

**Comment by vimla01** [comment]
> @vimla01 Thanks for addressing the comments and pushing the changes.
> 
> One small note for future PRs: please avoid resolving review comments yourself. It's usually better to leave them open so the reviewer can verify the change and resolve them after confirming everything looks good.

Got it @vishalm0509 , thanks! I'll leave review comments open next time.
I thought they should be resolved after addressing the changes.
